<h1>
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="docs/images/nf-core-multiplesequencealign_logo_dark.png">
    <img alt="nf-core/multiplesequencealign" src="docs/images/nf-core-multiplesequencealign_logo_light.png">
  </picture>
</h1>

[![GitHub Actions CI Status](https://github.com/nf-core/multiplesequencealign/actions/workflows/ci.yml/badge.svg)](https://github.com/nf-core/multiplesequencealign/actions/workflows/ci.yml)
[![GitHub Actions Linting Status](https://github.com/nf-core/multiplesequencealign/actions/workflows/linting.yml/badge.svg)](https://github.com/nf-core/multiplesequencealign/actions/workflows/linting.yml)[![AWS CI](https://img.shields.io/badge/CI%20tests-full%20size-FF9900?labelColor=000000&logo=Amazon%20AWS)](https://nf-co.re/multiplesequencealign/results)[![Cite with Zenodo](http://img.shields.io/badge/DOI-10.5281/zenodo.XXXXXXX-1073c8?labelColor=000000)](https://doi.org/10.5281/zenodo.XXXXXXX)
[![nf-test](https://img.shields.io/badge/unit_tests-nf--test-337ab7.svg)](https://www.nf-test.com)

[![Nextflow](https://img.shields.io/badge/nextflow%20DSL2-%E2%89%A524.04.1-23aa62.svg)](https://www.nextflow.io/)
[![run with conda](http://img.shields.io/badge/run%20with-conda-3EB049?labelColor=000000&logo=anaconda)](https://docs.conda.io/en/latest/)
[![run with docker](https://img.shields.io/badge/run%20with-docker-0db7ed?labelColor=000000&logo=docker)](https://www.docker.com/)
[![run with singularity](https://img.shields.io/badge/run%20with-singularity-1d355c.svg?labelColor=000000)](https://sylabs.io/docs/)
[![Launch on Seqera Platform](https://img.shields.io/badge/Launch%20%F0%9F%9A%80-Seqera%20Platform-%234256e7)](https://cloud.seqera.io/launch?pipeline=https://github.com/nf-core/multiplesequencealign)

[![Get help on Slack](http://img.shields.io/badge/slack-nf--core%20%23multiplesequencealign-4A154B?labelColor=000000&logo=slack)](https://nfcore.slack.com/channels/multiplesequencealign)[![Follow on Twitter](http://img.shields.io/badge/twitter-%40nf__core-1DA1F2?labelColor=000000&logo=twitter)](https://twitter.com/nf_core)[![Follow on Mastodon](https://img.shields.io/badge/mastodon-nf__core-6364ff?labelColor=FFFFFF&logo=mastodon)](https://mstdn.science/@nf_core)[![Watch on YouTube](http://img.shields.io/badge/youtube-nf--core-FF0000?labelColor=000000&logo=youtube)](https://www.youtube.com/c/nf-core)

## Introduction

**nf-core/multiplesequencealign** is a pipeline to deploy and systematically evaluate Multiple Sequence Alignment (MSA) methods.

The pipeline is built using [Nextflow](https://www.nextflow.io), a workflow tool to run tasks across multiple compute infrastructures in a very portable manner. It uses Docker/Singularity containers making installation trivial and results highly reproducible. The [Nextflow DSL2](https://www.nextflow.io/docs/latest/dsl2.html) implementation of this pipeline uses one container per process which makes it much easier to maintain and update software dependencies. Where possible, these processes have been submitted to and installed from [nf-core/modules](https://github.com/nf-core/modules) in order to make them available to all nf-core pipelines, and to everyone within the Nextflow community!

On release, automated continuous integration tests run the pipeline on a full-sized dataset on the AWS cloud infrastructure. This ensures that the pipeline runs on AWS, has sensible resource allocation defaults set to run on real-world datasets, and permits the persistent storage of results to benchmark between pipeline releases and other analysis sources.The results obtained from the full-sized test can be viewed on the [nf-core website](https://nf-co.re/proteinfold/results).

![Alt text](docs/images/nf-core-msa_metro_map.png?raw=true "nf-core-msa metro map")

1. **Input files summary**: (Optional) computation of summary statistics on the input fasta file, such as the average sequence similarity across the input sequences, their length, etc. Skip by `--skip_stats` as a parameter.
2. **Guide Tree**: (Optional) Renders a guide tree. 
3. **Align**: aligns the sequences.
4. **Evaluate**: (Optional) The obtained alignments are evaluated with different metrics: Sum Of Pairs (SoP), Total Column score (TC), iRMSD, Total Consistency Score (TCS), etc. Skip by passing `--skip_eval` as a parameter.
5. **Report**: Reports about the collected information of the runs are reported in a shiny app and a summary table in multiqc. Skip by passing `--skip_shiny` and `--skip_multiqc`.


### 1. INPUT FILES SUMMARY 

Summary information about the input fasta files are calculated. Skip by `--skip_stats`.

  1. Sequence similarity. Calculates pairwise and average sequence similarity is calculated using TCOFFEE. Activate with `--calc_sim` (default: false).
  2. General summary. Calculates the number and the average length of sequences. Activate with `--calc_seq_stats` (default: true).
  3. Extract plddt. If the structures were generated by AF2, plddt is extracted and reported. Activate with `--extract_plddt` (default: false).

### 2. GUIDE TREES

Guide trees define the order in which sequences and profiles are aligned and hold a crucial role in determining the final MSA accuracy. Tree rendering techniques most commonly rely on pairwise distances between sequences. 

> **Note**
> None of the below listed aligner needs an explicit definition of a guidetree: if they need one, they compute their own default guide tree. This explicit definition of a guide tree is available in case you want to test non-default combination of guide trees and aligner methods. 


Available GUIDE TREE methods (Optional):

- [CLUSTALO](http://clustal.org/omega/#Documentation)
- [FAMSA](https://github.com/refresh-bio/FAMSA)


### 3. ALIGN

Available ALIGN methods:

SEQUENCE-BASED (only require a fasta file as input):
- [CLUSTALO](http://clustal.org/omega/#Documentation) (accepts guide tree)
- [FAMSA](https://github.com/refresh-bio/FAMSA) (accepts guide tree)
- [KALIGN](https://github.com/TimoLassmann/kalign)
- [LEARNMSA](https://github.com/Gaius-Augustus/learnMSA)
- [MAFFT](https://mafft.cbrc.jp/alignment/server/index.html)
- [MAGUS](https://github.com/vlasmirnov/MAGUS) (accepts guide tree)
- [MUSCLE5](https://drive5.com/muscle5/manual/)
- [TCOFFEE](https://tcoffee.readthedocs.io/en/latest/index.html) (accepts guide tree)

SEQUENCE- and STUCTURE-BASED (require both fasta and structures as input):
- [3DCOFFEE](https://tcoffee.org/Projects/expresso/index.html) (accepts guide tree)

STRUCTURE-BASED (only require stuctures as input):
- [MTMALIGN](https://bio.tools/mtm-align) 

### 4. EVALUATE

Optionally, the produced MSAs can be evaluated. Skip with `--skip_eval`.

SEQUENCE-BASED (no extra input required): 
1. Calculate number of gaps and its average across sequences. Activate using `--calc_gaps` (default: true).

REFERENCE-BASED: 
The reference MSAs (see samplesheet) are be used to evaluate the quality of the produced MSA. 

2. Sum Of Pairs. Calculates the SP score using the [TCOFFEE](https://tcoffee.readthedocs.io/en/latest/index.html) implementation. Activate using `--calc_sp` (default: true).
3. Total column. Calculates the TC score [TCOFFEE](https://tcoffee.readthedocs.io/en/latest/index.html). Activate using `--calc_tc` (default: true).

STRUCTURE-BASED:
The provided structures (see samplesheet) are used to evaluate the quality of the alignment. 
4. iRMSD. Calculates the iRMSD using the [TCOFFEE](https://tcoffee.readthedocs.io/en/latest/index.html) implementation. Activate using `--calc_irmsd` (default: false).


### 5. REPORT

Finally, a summary table with all the computed statistics and evaluations is reported in MultiQC (Skip by `--skip_multiqc`).
Moreover, a shiny app is prepared with interactive summary plots. 
> [!WARNING]
> You will need to have [shiny](https://shiny.posit.co/py/) installed to run it! See [output documentation](https://nf-co.re/multiplesequencealign/output) for more infos.


## Usage

> [!NOTE]
> If you are new to Nextflow and nf-core, please refer to [this page](https://nf-co.re/docs/usage/installation) on how to set-up Nextflow. Make sure to [test your setup](https://nf-co.re/docs/usage/introduction#how-to-run-a-pipeline) with `-profile test` before running the workflow on actual data.


#### 1. SAMPLESHEET
The sample sheet defines the input data that the pipeline will process. 
It should look like this: 

`samplesheet.csv`:

```csv
id,fasta,reference,structures
seatoxin,seatoxin.fa,seatoxin-ref.fa,seatoxin_structures
toxin,toxin.fa,toxin-ref.fa,toxin_structures
```

Each row represents a set of sequences (in this case the seatoxin and toxin protein families) to be processed.

`id` is the name of the set of sequences. It can correspond to the protein family name or to an internal id. It should be unique.

The column `fasta` contains the path to the fasta file that contains the sequencest be aligned.

The column `reference` is optional and contains the path to the reference alignment. It is used for the reference-based evaluation steps. It can be left empty.

The column `structures` is also optional and contains the path to the folder that contains the protein structures for the sequences to be aligned. It is used for structural aligners and structure-based evaluation steps. It can be left empty.


> [!NOTE]
> You can have some samples with structures and/or references and some without. The pipeline will run the modules requiring structures/references only on the samples for which you have provided the required information and the others will be just skipped. 


#### 2. TOOLSHEET

The reason why we provide a toolsheet as input is that if the pipeline needs to be used as benchmarking framework, we need to test multiple arguments per module. This can be done by having multiple entries in the toolsheet per module with the multiple arguments to be tested. 

Each line of the toolsheet defines a combination of guide tree and multiple sequence aligner to run with the respective arguments to be used. 

It should look at foollows: 

`toolsheet.csv`:

```csv
tree,args_tree,aligner,args_aligner,
FAMSA, -gt upgma -medoidtree, FAMSA,
, ,TCOFFEE, 
FAMSA,,REGRESSIVE,
```
> [!NOTE]
> Each of the trees and aligners are available as standalones, so args_tree and args_aligner can be left empty if you are cool with the default settings of each method. Tree can also be left empty, and the default guide tree will be used for each aligner. 

> [!NOTE]
>  use the exact spelling as listed above!

`tree` is the tool used to build the tree. (optional)

Arguments to the tree tool can be provided using `args_tree`. Please refer to each tool's documentation. (optional)

The `aligner` column contains the tool to run the alignment. (optional)

Finally, the arguments to the aligner tool can be set by using the `args_alginer` column. (optional)


#### 3. RUN THE PIPELINE
Now, you can run the pipeline using:

```bash
nextflow run nf-core/multiplesequencealign \
   -profile test \
   --input samplesheet.csv \
   --tools toolsheet.csv \
   --outdir outdir
```

> [!WARNING]
> Please provide pipeline parameters via the CLI or Nextflow `-params-file` option. Custom config files including those provided by the `-c` Nextflow option can be used to provide any configuration _**except for parameters**_;
> see [docs](https://nf-co.re/usage/configuration#custom-configuration-files).

For more details and further functionality, please refer to the [usage documentation](https://nf-co.re/multiplesequencealign/usage) and the [parameter documentation](https://nf-co.re/multiplesequencealign/parameters).

## Pipeline output

To see the results of an example test run with a full size dataset refer to the [results](https://nf-co.re/multiplesequencealign/results) tab on the nf-core website pipeline page.
For more details about the output files and reports, please refer to the
[output documentation](https://nf-co.re/multiplesequencealign/output).

## Credits

nf-core/multiplesequencealign was originally written by Luisa Santus ([@luisas](https://github.com/luisas)) and Jose Espinosa-Carrasco ([@JoseEspinosa](https://github.com/JoseEspinosa)) from The Comparative Bioinformatics Group at The Centre for Genomic Regulation, Spain.

The following people have significantly contributed to the development of the pipeline and its modules: Leon Rauschning ([@lrauschning](https://github.com/lrauschning)), Alessio Vignoli ([@alessiovignoli](https://github.com/alessiovignoli)) and Leila Mansouri ([@l-mansouri](https://github.com/l-mansouri)).

## Contributions and Support

If you would like to contribute to this pipeline, please see the [contributing guidelines](.github/CONTRIBUTING.md).

For further information or help, don't hesitate to get in touch on the [Slack `#multiplesequencealign` channel](https://nfcore.slack.com/channels/multiplesequencealign) (you can join with [this invite](https://nf-co.re/join/slack)).

## Citations

<!-- TODO nf-core: Add citation for pipeline after first release. Uncomment lines below and update Zenodo doi and badge at the top of this file. -->
<!-- If you use nf-core/multiplesequencealign for your analysis, please cite it using the following doi: [10.5281/zenodo.XXXXXX](https://doi.org/10.5281/zenodo.XXXXXX) -->

An extensive list of references for the tools used by the pipeline can be found in the [`CITATIONS.md`](CITATIONS.md) file.

You can cite the `nf-core` publication as follows:

> **The nf-core framework for community-curated bioinformatics pipelines.**
>
> Philip Ewels, Alexander Peltzer, Sven Fillinger, Harshil Patel, Johannes Alneberg, Andreas Wilm, Maxime Ulysse Garcia, Paolo Di Tommaso & Sven Nahnsen.
>
> _Nat Biotechnol._ 2020 Feb 13. doi: [10.1038/s41587-020-0439-x](https://dx.doi.org/10.1038/s41587-020-0439-x).
